ldpd/ldpd.c

```diff
@@ -415,7 +415,7 @@ ldpd_shutdown(void)
 			log_warnx("%s terminated; signal %d",
 			    (pid == lde_pid) ? "label decision engine" :
 			    "ldp engine", WTERMSIG(status));
-	} while (pid != -1 || (pid == -1 && errno == EINTR));
+	} while (pid != -1 || errno == EINTR);
```

Comment:
I think the second pid check is indeed redundant, but IMHO the original code is easier to read and understand. @rzalamena any opinion on this?

Reply:
I don't think this is a evident correction and the proposed solution is wrong. Here are some points to take into account:

1. It assumes `errno` is cleaned each iteration; What happens if `errno == EINTR` and `pid != -1`?

2. The current way is cleaner than the proposed solution, whoever I also think the current way is still not obvious.

From what I could understand `pid` is being used as state to determine whether should we continue or not the loop, and the condition to stop is: `pid == -1 && errno != EINTR`.

3. The current code is actually a boiler plate of daemon code used in many OpenBSD daemons, it might have more experience than meet the eyes (perhaps some *BSD specifics).

---

I'm going to try be corageous and propose a solution: here is my suggestion of code, it may or may not be a cleaner alternative:

```c
while (true) {
        /* Wait for child process. */
        pid = wait(&status);
        if (pid == -1) {
                /* We got interrupted, try again. */
                if (errno == EINTR)
                        continue;
                /* No more processes were found. */
                if (errno == ECHILD)
                        break;

                /* Unhandled errno condition. */
                fatal("wait");
                /* UNREACHABLE */
        }

        /* We found something, lets announce it. */
        if (WIFISIGNALED(status))
                log_warnx("%s terminated; signal %d",
                          (pid == lde_pid) ? "label decision engine" :
                          "ldp engine", WTERMSIG(status));

        /* Repeat until there are no more child processes. */
}
```

Feel free to use it and remove comments if you want, I have not tested it though.